### DISCUSSION TITLE
pass complex data structures to python through PyCom

### DISCUSSION THREAD

**Discussion by KeanDequeant** [the-anylogic-company/AnyLogic-Pypeline#57] [Q&A] [answered]
Hello! 
First of all, a huge thank you for Pypeline! 
I have been using it for a year now as part of my work (with the Anylogic professional licence), and being able to execute our production code from within Anylogic is a game changer -> that's what allowed us to build our first factory digital twin, and it was a big factor in our decision to prefer Anylogic over the competition.  
However, one limiting factor that I see is on the ability to transfer complex data structures. As our production code is mainly in python, a lot of our data structures are in table format (dataframes), but for now the only way I've found to store that information properly in Anylogic is to create an agent for every row of each dataframe, which can get very big with inventory data (hundreds of thousands). 
I've tried using complex data structures such as nested HashMap of ArrayList, or ArrayList of HashMap... but the toJson function of pyCom doesn't work for those types of imbedded structures. 
I also thought of using the database tables from Anylogic, but I wasn't able to convert that data to json to pass to python.. 
Do you have any thoughts, recommendation on this? I.e. what would be the recommended way of passing a lot of table structured data from Python to Anylogic, (and back)?

**Comment by t-wolfeadam** [comment]
Hi! Thanks for using it and I'm very glad it could be of help 😄! 

I absolutely feel that there is a better method than turning rows into agents, some initial ideas for receiving it:
- at the very least, turning rows into simple Java classes will reduce memory overhead and some performance of custom agents (define the classes with `public` class variables for each column name)
- assuming you mean pandas' dataframe, you could convert it via `to_json`, likely using 'records' or 'index' orientation is simplest. `fromJson` should work for this but keep in mind Java's typing means you'll need to type cast the Object types to whatever their underlying type is. If I'm misunderstanding why it doesn't work, please clarify
- unsure of the overhead but pandas' `to_excel` or `to_sql` might also work, importing it into the AL DB (but it'll need to be setup in advance to accept the expected formats)

As far as sending it back to Python, the latest version (with the JSONifier object built-in) actually has JSON serializers for the database itself and individual tables. Below is an example. It may not be in the desired format, but you can either process that in Python or override the serializer. I'll try to think of some more possibilities (but I'm out until next week)!

![image](https://github.com/t-wolfeadam/AnyLogic-Pypeline/assets/40703071/d852e6c6-09c0-40b6-94f7-409cf0d4206c)

**Comment by KeanDequeant** [comment]
Ah great! 
Yes my problem was more in passing data back from Anylogic to python (and therefore how to store the data in Anylogic), but what you've shown with the Anylogic tables looks very promising! Apologies I didn't think of looking at the latest releases of Pypeline, indeed I was using version 1.7.1. 
-> so with the latest version of Pypeline, is it as easy as doing PyCom.toJson(table_name) ? 
I will try that as soon as I have a bit of time, thanks a lot for the support!

**Reply by t-wolfeadam** [reply]
> is it as easy as doing PyCom.toJson(table_name) ?

It should be but I just noticed the latest version only outputs the tables' metadata (column info and constraints, not data). So currently you need to either do it for the whole database (e.g., `pyCom.toJson(getDatabaseConnection())`) or after converting the table to a list of records (e.g., `selectFrom(main_table).list()`), but you also have to map from the Tuple object since it's not (currently) serializable (e.g., `pyCom.toJson( selectFrom(main_table).list().stream().map(t -> t.toArray()).toArray(Object[]::new) )`). 

In case it isn't clear, here's an example: [Pypeline DB JSONify Mini Demo.zip](https://github.com/t-wolfeadam/AnyLogic-Pypeline/files/12423178/Pypeline.DB.JSONify.Mini.Demo.zip)

**Reply by t-wolfeadam** [reply]
Disregard the previous comment, I fixed the issues 🙂 
If you download/update to [the current version (>=1.9.3) ](https://github.com/t-wolfeadam/AnyLogic-Pypeline/releases), you can now jsonifiy single tables or their contents directly.

![image](https://github.com/t-wolfeadam/AnyLogic-Pypeline/assets/40703071/6d4c34bc-4a56-4d97-808e-de8f76942c14)

**Reply by KeanDequeant** [reply]
amazing thank you so much! I will try it as soon as I have time to work on this again. This is going to be a big game changer :)

**Comment by KeanDequeant** [comment]
sorry I'm only coming back to that problem now! The way Anylogic -> Python works fine, thanks a lot for that! 

The bit that blocks more is passing data from Python to Anylogic, and store that data in one of the Anylogic database table. So far what I'm doing seems to be a bit of a workaround, as I am storing the data I get from python into a collection (using PyCom.fromJson), and then I iterate over that collection to insert rows in the Anylogic database table. 
Is there maybe a more direct way of doing that? 

Thanks!

**Reply by t-wolfeadam** [reply, marked as answer]
No worries, great to hear it's working out so far!
 
There's no way to directly dump to a table at the moment. The easiest way to do it is as you're doing - just to iterate over the values. Other possible ways I can think of:

- have your Python code generate an sql string rather than using the QuerySQL api (reference: [anylogic help](https://anylogic.help/anylogic/connectivity/querying.html#insert)); this obviously requires the python code to be in sync with the structure of the model
- connect to the database directly from Python (read more in [this SO post](https://stackoverflow.com/a/42758347)) - though, technically this should only be done with the model closed

**Reply by KeanDequeant** [reply]
nice thanks! yeah I didn't think of writing an sql string straight from python -> that's probably the most efficient!
